docs/architecture.md

```diff
@@ -0,0 +1,50 @@
+
+Upon startup, engine would login as all bots individually and maintain a pool of matrix client instances and starts the `sync` process with the homeserver, giving each bot the chance of reacting to events as they come in. It also loads the triggers, workflows and workflow steps that are defined in the database. Do note that TOML file is only parsed once & imported at startup and then everything happens based on the data inside the database. Its only when the program starts again, that TOML file is reimported. In future, we would implement signalling the program to reload TOML file without requiring a reload of the main program itself.
+
+When triggers are loaded, it starts the monitoring process of defined triggers. For `webhook` variety of triggers, we start a single webhooks listener server, which handles all incoming HTTP requests from outside services. All endpoints share a common prefix `webhooks-listener`. For `poller` variety of triggers, it invokes setup mechanism of these triggers, based on which they can keep polling. This isn't well-built yet, just the skeleton of the mechanism exist.
+
+More variety of triggers are planned such as:
+- Matrix based events (commands invoked, emoji reactions etc)
+- Schedule based (Cron)
+
+When workflow steps are loaded, they are just queued up in their specified order within a particular workflow and await start of the workflow. When a workflow starts, it may or may not have a payload to pass to the first workflow step. Every workflow step would accept the payload from the previous workflow step and passes it forward, with any modification it chooses to make to it.
+
+Each trigger and workflow step carries additional meta information based on their variety.
+
+## What other varieties of workflow steps are planned?
+
+Hard to put an exhaustive list, as we would build what we need first. Some are:
+
+- Ping an external endpoint with payload data
+- Query API to add more data to payload data
+- DM a certain user
+- Ask questions to a group of users in a DM and aggregate those answers & post in a matrix room. `Stand up meetings`
+- Send email
+
+## What else is planned for the future?
+
+### Keeping tabs on who's online
+
+We have a polyglots command in Automattic, which when invoked can help you find someone who speaks a certain language and is online right now. Supporting such a command would require knowing who is online and this is probably best done by `neurobot` by maintaining a "online users" list, which can simply be utilised by a workflow step.
```

Comment:
Considering this is a public repo, I'm wondering should we move this to an issue on our board instead?